apollo-federation/src/operation/simplify.rs

```diff
@@ -0,0 +1,532 @@
+                schema,
+                RebaseErrorHandlingOption::ThrowError,
+            )?;
+            Ok(Some(SelectionOrSet::Selection(Selection::InlineFragment(
+                Arc::new(InlineFragmentSelection::new(
+                    rebased_inline_fragment,
+                    rebased_selection_set,
+                )),
+            ))))
+        } else {
+            unreachable!("We should always be able to either rebase the inline fragment OR throw an exception");
+        }
+    }
+}
+
+impl SelectionSet {
+    /// Simplify this selection set in the context of the provided `parent_type`.
+    ///
+    /// This removes unnecessary/redundant inline fragments, so that for instance, with a schema:
+    /// ```graphql
+    /// type Query {
+    ///   t1: T1
+    ///   i: I
+    /// }
+    ///
+    /// interface I {
+    ///   id: ID!
+    /// }
+    ///
+    /// type T1 implements I {
+    ///   id: ID!
+    ///   v1: Int
+    /// }
+    ///
+    /// type T2 implements I {
+    ///   id: ID!
+    ///   v2: Int
+    /// }
+    /// ```
+    /// We can perform following simplification:
+    /// ```graphql
+    /// flatten_unnecessary_fragments({
+    ///   t1 {
+    ///     ... on I {
+    ///       id
+    ///     }
+    ///   }
+    ///   i {
+    ///     ... on T1 {
+    ///       ... on I {
+    ///         ... on T1 {
+    ///           v1
+    ///         }
+    ///         ... on T2 {
+    ///           v2
+    ///         }
+    ///       }
+    ///     }
+    ///     ... on T2 {
+    ///       ... on I {
+    ///         id
+    ///       }
+    ///     }
+    ///   }
+    /// }) === {
+    ///   t1 {
+    ///     id
+    ///   }
+    ///   i {
+    ///     ... on T1 {
+    ///       v1
+    ///     }
+    ///     ... on T2 {
+    ///       id
+    ///     }
+    ///   }
+    /// }
+    /// ```
+    ///
+    /// For this operation to be valid (to not throw), `parent_type` must be such that every field selection in
+    /// this selection set is such that its type position intersects with passed `parent_type` (there is no limitation
+    /// on the fragment selections, though any fragment selections whose condition do not intersects `parent_type`
+    /// will be discarded). Note that `self.flatten_unnecessary_fragments(self.type_condition)` is always valid and useful, but it is
+    /// also possible to pass a `parent_type` that is more "restrictive" than the selection current type position
+    /// (as long as the top-level fields of this selection set can be rebased on that type).
+    ///
+    // PORT_NOTE: this is now module-private, because it looks like it *can* be. If some place
+    // outside this module *does* need it, feel free to mark it pub(crate).
+    // PORT_NOTE: in JS, this was called "normalize".
+    // PORT_NOTE: in JS, this had a `recursive: false` flag, which would only apply the
+    // simplification at the top level. This appears to be unused.
+    pub(super) fn flatten_unnecessary_fragments(
```

Comment:
maybe this should just be `flatten`?

Reply:
when we talked about it about a month ago I got the sense that people preferred being specific about it affecting fragments